src/parser/core/modules/BuffWindow.tsx

```diff
@@ -0,0 +1,329 @@
+		if ( lastBuffWindow && lastBuffWindow.end == null ) {
+			return lastBuffWindow
+		}
+		return undefined
+	}
+
+	protected init() {
+		this.addHook('cast', {by: 'player'}, this.onCast)
+		this.addHook('applybuff', {by: 'player'}, this.onApplyBuff)
+		this.addHook('removebuff', {by: 'player'}, this.onRemoveBuff)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onCast(event: CastEvent) {
+		const action: ActionData | undefined = getDataBy(ACTIONS, 'id', event.ability.guid)
+
+		if (!action || action.autoAttack) {
+			// Disregard auto attacks for tracking rotations / events during buff windows
+			return
+		}
+
+		if (this.activeBuffWindow) {
+			this.activeBuffWindow.rotation.push(event)
+		}
+	}
+
+	private onApplyBuff(event: BuffEvent) {
+		if (!this.buffStatus || event.ability.guid !== this.buffStatus.id) {
+			return
+		}
+
+		this.startNewBuffWindow(event.timestamp)
+	}
+
+	private startNewBuffWindow(startTime: number) {
+		this.buffWindows.push(new BuffWindowState(startTime))
+	}
+
+	private onRemoveBuff(event: BuffEvent) {
+		if (!this.buffStatus || event.ability.guid !== this.buffStatus.id) {
+			return
+		}
+
+		if (this.activeBuffWindow) {
+			this.activeBuffWindow.end = event.timestamp
+		}
+	}
+
+	// For consumers that have the same number of expected GCDs per window, this will use the expectedPerWindow property
+	//   on expectedGCDs as the baseline.  This method can be overridden if the logic of expected GCDs per window is variable
+	protected getBaselineExpectedGCDs(buffWindow: BuffWindowState): number {
+		if ( this.expectedGCDs ) {
+			return this.expectedGCDs.expectedPerWindow
+		}
+		return 0
+	}
+
+	// Override point for class-specific rushing logic per BuffWindow - default no effect
+	// Return a positive number to increase expected GCDs for this window, and a negative number to decrease
+	protected changeExpectedGCDsClassLogic(buffWindow: BuffWindowState): number {
+		return 0
+	}
+
+	// Rushing handling for end of fight rushing.  Can be overridden if class rules for end of fight rushing vary
+	protected reduceExpectedGCDsEndOfFight(buffWindow: BuffWindowState): number {
+		if ( this.buffStatus && this.buffStatus.duration ) {
+			// Check to see if this window is rushing due to end of fight - reduce expected GCDs accordingly
+			const windowDurationMillis = this.buffStatus.duration * 1000
+			const fightTimeRemaining = this.parser.fight.end_time - buffWindow.start
+
+			if (windowDurationMillis >= fightTimeRemaining) {
+				const gcdEstimate = this.globalCooldown.getEstimate()
+				return Math.ceil((windowDurationMillis - fightTimeRemaining) / gcdEstimate)
+			}
+		}
+
+		// Default: no rushing reduction
+		return 0
+	}
+
+	// For consumers that have tracked cooldowns that expect the same number of that cooldown per window, this will use the
+	//   expectedPerWindow property on that cooldown as the baseline.  This method can be overridden if the logic of
+	//   expected tracked cooldowns per window is variable
```

Comment:
For comments about a method or prop, _especially_ those used by other files, consider using TSDoc comments (`/** ... */`), as they'll show up in intellisense without needing to check the source of the imported file.